src/redux-middleware/pushQueue.ts

```diff
@@ -139,13 +133,54 @@ const flushPushQueue = (): Thunk<Promise<void>> => async (dispatch, getState) =>
 
   if (pushQueue.length === 0) return Promise.resolve()
 
-  dispatch(flushDeletes())
-  dispatch(flushEdits())
-  dispatch(flushMoves())
+  /**
+   * Returns updated thoughtIndexUpdates using the latest state after pulling pending lexemes.
+   */
+  const getSyncedThoughtsUpdate = (batch: PushBatch) => Object.keys(batch.thoughtIndexUpdates).reduce<Index<Lexeme | null>>((acc, thoughtId) => {
+
+    const updatedState = getState()
+
+    // Return early if this particular lexeme was not pending
+    const hadPendingLexeme = batch.pendingLexemes?.[thoughtId]
+    if (!hadPendingLexeme) return acc
+
+    // Note: Assuming thoughtIndexUpdates entry with null value won't occur, as they will have associated lexeme already pulled.
```

Comment:
Could you clarify "they will have associated lexeme already pulled"? Are null updates somehow removed before this point?

Reply:
null updates are caused by thought deletion. For thought deletion, we already need to pull the associated lexemes and parent (from the pullQueue). So after null updates won't run into a state where the associated lexeme is not available. Please correct me if I am wrong to assume this.

Reply:
An edit consists of a null update + a new or modified update. There will be no conflicting Lexemes from a null update, but that doesn't mean there are no null updates. 

Or do you mean we know there will be no null updates because of the check on L168 of pushQueue?

Reply:
`pendingLexemes` are the `thoughtIndexUpdates` that do not have an entry in the `thoughtIndex`. For all null `thoughtIndexUpdates`, there would always be an entry in the `thoughtIndex` as they are pulled by the `pullQueue` before the delete operation.

Reply:
Is it the conditional above, `if (!hadPendingLexeme) return acc`, that ensures this? If not, I give up 😅

Reply:
> Is it the conditional above, `if (!hadPendingLexeme) return acc`, that ensures this? If not, I give up 😅

Yes. It checks if the update was pending or not.